### PR TITLE
store/tikv/mock-tikv: add coprocessor support.

### PR DESCRIPTION
**Pull request by coocood**
As mockTiKV currently is much slower than `localstore` and do not support store data on disk, currently only `executor` package is set to use mockTikv. `localstore` should still be used for a while.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @disksing PTAL

File: store/tikv/kv.go [hunk 1]
```diff
@@ -88,6 +89,21 @@ func newTikvStore(uuid string, pdClient pd.Client, factory ClientFactory) *tikvS
...
 }
 
+// NewMockTikvStore creates a mocked tikv store.
+func NewMockTikvStore() kv.Storage {
+	cluster := mocktikv.NewCluster()
+	mocktikv.BootstrapWithSingleStore(cluster)
+	mvccStore := mocktikv.NewMvccStore()
+	clientFactory := mockClientFactory(cluster, mvccStore)
+	return newTikvStore("mock-tikv-store", mocktikv.NewPDClient(cluster), clientFactory)
+}
```
Comment (disksing) [inline store/tikv/kv.go]: Maybe better to put this into a `_test.go` file if it's only used in test.

File: store/tikv/mock-tikv/mvcc.go [hunk 2]
```diff
@@ -241,6 +241,40 @@ func (s *MvccStore) Scan(startKey, endKey []byte, limit int, startTS uint64) []P
...
+			pair = Pair{Key: k, Value: val, Err: err}
+			return false
+		}
+		return true
+	}
+	if reverse {
+		s.tree.DescendLessOrEqual(newEntry(key), iterator)
+	} else {
+		s.tree.AscendGreaterOrEqual(newEntry(key), iterator)
+	}
```
Comment (disksing) [inline store/tikv/mock-tikv/mvcc.go]: Access the tree without lock here. I suggest just using `scan 1` like tikv-server.

File: store/tikv/kv.go (hunk 1 above)
Comment (coocood) [inline store/tikv/kv.go]: We may use it in tidb-server later.

**Comment by coocood** [comment]
@disksing PTAL

File: store/tikv/coprocessor.go [hunk 3]
```diff
@@ -337,7 +338,6 @@ func (it *copIterator) handleTask(task *copTask) (*coprocessor.Response, error)
 		}
 		it.mu.Lock()
 		defer it.mu.Unlock()
-		task.status = taskDone
```
Comment (disksing) [inline store/tikv/coprocessor.go]: Above 2 lines can be removed.

**Comment by coocood** [comment]
@disksing @shenli @zimulala PTAL

File: executor/executor_test.go [hunk 4]
```diff
@@ -42,9 +44,16 @@ type testSuite struct {
 }
 
 func (s *testSuite) SetUpSuite(c *C) {
-	store, err := tidb.NewStore("memory://test/test")
-	c.Assert(err, IsNil)
-	s.store = store
+	useMockTikv := true
```
Comment (zimulala) [inline executor/executor_test.go]: Here `useMockTikv` value is always `true`, why use `if` ?

File: store/tikv/mock-tikv/mvcc.go [hunk 5]
```diff
@@ -213,7 +213,7 @@ func regionContains(startKey []byte, endKey []byte, key []byte) bool {
 		(bytes.Compare(key, endKey) < 0 || len(endKey) == 0)
 }
 
-// Scan reads up to limit numbers of Pairs from a key.
+// Scan reads up to limit numbers of Pairs that greater or equal to startKey and less than endKey.
```
Comment (zimulala) [inline store/tikv/mock-tikv/mvcc.go]: greater -> greater than ?

File: store/tikv/mock-tikv/mvcc.go [hunk 6]
```diff
@@ -241,6 +241,38 @@ func (s *MvccStore) Scan(startKey, endKey []byte, limit int, startTS uint64) []P
 	return pairs
 }
 
+// ReverseScan reads up to limit numbers of Pairs that greater or equal to startKey and less than endKey
```
Comment (zimulala) [inline store/tikv/mock-tikv/mvcc.go]: numbers of -> the numbers of ?
greater -> greater than ?

**Comment by disksing** [comment]
LGTM

File: xapi/xapi.go [hunk 7]
```diff
@@ -146,12 +147,14 @@ func composeRequest(req *tipb.SelectRequest, concurrency int) (*kv.Request, erro
...
 			kvReq.Desc = *req.OrderBy[0].Desc
-			kvReq.KeepOrder = true
 		}
 	} else {
 		kvReq.Tp = kv.ReqTypeSelect
 		tid := req.GetTableInfo().GetTableId()
 		kvReq.KeyRanges = tablecodec.EncodeTableRanges(tid, req.Ranges)
+		if req.OrderBy != nil {
+			kvReq.Desc = *req.OrderBy[0].Desc
+		}
```
Comment (zimulala) [inline xapi/xapi.go]: Line 155-157 and Line 148-150 are the same. So line 155-157 can move behind the line 158?

File: executor/executor_test.go (hunk 4 above)
Comment (coocood) [inline executor/executor_test.go]: It is easier to change running storage by just change 'true' to 'false'.

**Comment by coocood** [comment]
@zimulala PTAL

File: store/tikv/mock-tikv/cop_handler.go [hunk 8]
```diff
@@ -0,0 +1,474 @@
+package mocktikv
```
Comment (shenli) [inline store/tikv/mock-tikv/cop_handler.go]: Miss license.

File: store/tikv/mock-tikv/cop_handler.go [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+
+func (h *rpcHandler) getRowsFromSelectReq(ctx *selectContext) ([]*tipb.Row, error) {
+	kvRanges, desc := h.extractKVRanges(ctx.sel)
+	var rows []*tipb.Row
+	limit := int64(-1)
+	if ctx.sel.Limit != nil {
+		limit = ctx.sel.GetLimit()
+	}
+	for _, ran := range kvRanges {
+		if limit == 0 {
```
Comment (shenli) [inline store/tikv/mock-tikv/cop_handler.go]: limit <= 0 ?

Comment (coocood) [inline store/tikv/mock-tikv/cop_handler.go]: Negative limit means unlimited.

File: store/tikv/mock-tikv/cop_handler.go [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+	var d types.Datum
+	d.SetInt64(handle)
+	var err error
+	row.Handle, err = codec.EncodeValue(nil, d)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	for _, col := range columns {
+		if *col.PkHandle {
+			row.Data = append(row.Data, row.Handle...)
```
Comment (shenli) [inline store/tikv/mock-tikv/cop_handler.go]: Check Unsigned Flag.

File: store/tikv/mock-tikv/cop_handler.go [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+		} else {
+			colID := col.GetColumnId()
+			if ctx.whereColumns[colID] != nil {
+				// The column is saved in evaluator, use it directly.
+				datum := ctx.eval.Row[colID]
+				row.Data, err = codec.EncodeValue(row.Data, datum)
+				if err != nil {
```
Comment (shenli) [inline store/tikv/mock-tikv/cop_handler.go]: err -> err1

File: store/tikv/mock-tikv/cop_handler.go [hunk 12]
```diff
@@ -0,0 +1,487 @@
...
+}
+
+func (h *rpcHandler) evalWhereForRow(ctx *selectContext, handle int64) (bool, error) {
+	if ctx.sel.Where == nil {
+		return true, nil
+	}
+	tid := ctx.sel.TableInfo.GetTableId()
+	for colID, col := range ctx.whereColumns {
+		if col.GetPkHandle() {
+			ctx.eval.Row[colID] = types.NewIntDatum(handle)
```
Comment (shenli) [inline store/tikv/mock-tikv/cop_handler.go]: Check unsigned flag?

File: store/tikv/mock-tikv/cop_handler.go (hunk 11 above)
Comment (coocood) [inline store/tikv/mock-tikv/cop_handler.go]: This doesn't compile with error message `non-name row.Data on left side of :=`.

**Comment by coocood** [comment]
@BusyJay see the latest commit https://github.com/pingcap/tidb/pull/1230/commits/8ed1fb047169ac71caa392f9badffcd398e318dc

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**